cli_tools/osconfig_agent/patch_window.go

```diff
@@ -15,26 +15,306 @@
 package main
 
 import (
+	"encoding/json"
 	"errors"
 	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"sync"
 	"time"
 
 	osconfigpb "github.com/GoogleCloudPlatform/compute-image-tools/cli_tools/osconfig_agent/_internal/gapi-cloud-osconfig-go/google.golang.org/genproto/googleapis/cloud/osconfig/v1alpha1"
+	"github.com/golang/protobuf/jsonpb"
 	"google.golang.org/genproto/googleapis/type/dayofweek"
 	"google.golang.org/genproto/googleapis/type/timeofday"
 )
 
+const state = "osconfig_patch.state"
+
+var (
+	aw activeWindows
+	rc chan *patchWindow
+)
+
+func patchInit() {
+	aw.windows = map[string]*patchWindow{}
+	rc = make(chan *patchWindow)
+
+	// Load current patch state off disk.
+	pw, err := loadState(state)
+	if err != nil {
+		log.Println("ERROR:", err)
+	} else if pw != nil && pw.Name != "" {
+		pw.register()
+	}
+
+	// Start the patch runner goroutine.
+	go patchRunner()
+}
+
+type activeWindows struct {
+	windows map[string]*patchWindow
+	mx      sync.Mutex
+}
+
+func (a *activeWindows) delete(name string) {
+	a.mx.Lock()
+	delete(a.windows, name)
+	a.mx.Unlock()
+}
+
+func (a *activeWindows) add(name string, w *patchWindow) {
+	a.mx.Lock()
+	a.windows[name] = w
+	a.mx.Unlock()
+}
+
+type patchPolicy struct {
+	*osconfigpb.LookupConfigsResponse_EffectivePatchPolicy
+}
+
+// MarshalJSON marchals a patchPolicy using jsonpb.
+func (p *patchPolicy) MarshalJSON() ([]byte, error) {
+	m := jsonpb.Marshaler{}
+	s, err := m.MarshalToString(p)
+	if err != nil {
+		return nil, err
+	}
+	return []byte(s), nil
+}
+
+// UnmarshalJSON unmarshals apatchPolicy using jsonpb.
+func (p *patchPolicy) UnmarshalJSON(b []byte) error {
+	return jsonpb.UnmarshalString(string(b), p)
+}
+
 type patchWindow struct {
-	start, end time.Time
+	Name   string
+	Policy *patchPolicy
+
+	Start, End time.Time
+
+	t      *time.Timer
+	cancel chan struct{}
+	mx     sync.RWMutex
```

Comment:
Seeing a mutext on the patch window has me wondering. If every patch window has its own mutex, what do they use to avoid conflicts with each other? Also, if every patch window needs a mutex, that seems like a lot of possible contention and threads/agents that it might be hard to avoid race condition bugs. I still wonder if it's not possible to reduce the amount of concurrency to simplify what is going on.

I'm still not up to speed with Go's concurrency semantics, but in a traditional threaded model, I'd see this agent having 2 threads - 1 for periodically calling the service and 1 for running OsConfig and PatchPolicy updates. Yes, there's some lock contention between the two, but for the most part, they wait and operate independently. And the shared state can be minimized and accessed in a small and controlled way.